src/audio/tdfb/tdfb_ipc4.c

```diff
@@ -0,0 +1,198 @@
+					 CTRL_INDEX_AZIMUTH_ESTIMATE);
+	if (!cd->msg) {
+		comp_err(mod->dev, "Failed to initialize control notification.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+void tdfb_send_notification_common(struct processing_module *mod)
+{
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	tdfb_send_notification(cd->msg, cd->az_value_estimate);
+}
+
+int tdfb_get_config_common(struct processing_module *mod,
+			   uint32_t param_id, uint32_t *data_offset_size,
+			   uint8_t *fragment, size_t fragment_size)
+{
+	comp_err(mod->dev, "tdfb_get_config_common(), Not supported, should not happen");
+	return -EINVAL;
+}
+
+static int tdfb_cmd_enum_set(struct sof_ipc4_control_msg_payload *ctl, struct tdfb_comp_data *cd)
+{
+	if (ctl->num_elems != 1)
+		return -EINVAL;
+
+	if (ctl->chanv[0].value > SOF_TDFB_MAX_ANGLES)
+		return -EINVAL;
+
+	switch (ctl->id) {
+	case CTRL_INDEX_AZIMUTH:
+		cd->az_value = ctl->chanv[0].value;
+		cd->update = true;
+		break;
+	case CTRL_INDEX_AZIMUTH_ESTIMATE:
+		cd->az_value_estimate = ctl->chanv[0].value;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int tdfb_cmd_switch_set(struct sof_ipc4_control_msg_payload *ctl, struct tdfb_comp_data *cd)
+{
+	if (ctl->num_elems != 1)
+		return -EINVAL;
+
+	switch (ctl->id) {
+	case CTRL_INDEX_PROCESS:
+		cd->beam_on = ctl->chanv[0].value;
+		cd->update = true;
+		break;
+	case CTRL_INDEX_DIRECTION:
+		cd->direction_updates = ctl->chanv[0].value;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int tdfb_set_config_common(struct processing_module *mod, uint32_t param_id,
+			   enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+			   const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+			   size_t response_size)
+{
+	struct sof_ipc4_control_msg_payload *ctl = (struct sof_ipc4_control_msg_payload *)fragment;
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	switch (param_id) {
+	case SOF_IPC4_SWITCH_CONTROL_PARAM_ID:
+		comp_info(mod->dev, "SOF_IPC4_SWITCH_CONTROL_PARAM_ID id = %d, num_elems = %d",
+			  ctl->id, ctl->num_elems);
+		return tdfb_cmd_switch_set(ctl, cd);
+
+	case SOF_IPC4_ENUM_CONTROL_PARAM_ID:
+		comp_info(mod->dev, "SOF_IPC4_ENUM_CONTROL_PARAM_ID id = %d, num_elems = %d",
+			  ctl->id, ctl->num_elems);
+		return tdfb_cmd_enum_set(ctl, cd);
+	default:
+		comp_info(mod->dev, "tdfb_set_config(), binary");
+		return comp_data_blob_set(cd->model_handler, pos, data_offset_size,
+					  fragment, fragment_size);
+	}
+}
+
+void tdfb_params(struct processing_module *mod)
+{
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct comp_buffer *sinkb, *sourceb;
+	struct comp_dev *dev = mod->dev;
+
+	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
+	component_set_nearest_period_frames(dev, params->rate);
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
```

Comment:
I realized here's similar issue as there was with my work with rtc aec. This is similar as original, but it's not handling correctly a typical case where source has more channels than sink. Please add a reminder /* TODO:  */ note for it. I can fix it later when I have suitable topology to use for test.

Reply:
@lkoenig FYI

Reply:
The merged fix for it is here: https://github.com/thesofproject/sof/pull/8452